feat: allow user to deposit any amount and set expiration date accordingly

pscott opened the pull request:
- refactor the expiration time computation to allow the user to deposit more than the yearly / monthly price, and update the expiration date based on the monthly / yearly price per second.
- Remove support for base network

bonustrack commented on src/writers.ts:
```diff
@@ -32,10 +94,29 @@ export function createEvmWriters(indexerName: string) {
+    // We do not save space entity if the computed expiration date is null
+    if (expirationDate !== null) {
+      space.turbo_expiration_timestamp = expirationDate.getTime() / MILLISECONDS; // Divide by 1000 to convert to seconds
+      space.turbo_expiration_date = expirationDate.toDateString();
```
We dont really need this value, we can take it from the timestamp

bonustrack commented on src/writers.ts:
```diff
@@ -1,12 +1,74 @@
+  }
+
+  // If the space already has an expiration date, use it as the current expiration date
+  const currentexpirationTimestamp = space.turbo_expiration_timestamp
```
```suggestion
  const currentExpirationTimestamp = space.turbo_expiration_timestamp
```

wa0x6e commented on src/writers.ts:
```diff
@@ -1,12 +1,78 @@
+const MILLISECONDS = 1000;
+const DECIMALS = 1e6; // USDC and USDT both have 6 decimals
+
+const TURBO_MONTHLY_PRICE = 1 * DECIMALS; // 600 USDC | USDT
```
I would prefer to have this also customizable via environment variables, so we could override it on testnet with smaller amount, without updating the code

pscott replied:
3958ae823fe67b66203f2fd2934deabbb59a63ae good idea! done

pscott replied:
didn't put the `decimals` as env vars because at the end of the day it's just a matter of adding `0` to your tests or not, should be easy. But changed the yearly and monthly to env vars

wa0x6e commented:
I am getting this error 

```
[1744074515668] INFO: found contract event
    component: "container"
    indexer: "sep"
    contract: "0xe40bfeb5a3014c9b98597088ca71eccdc27ca410"
    event: "PaymentReceived(address,address,uint256,string)"
    handlerFn: "handlePaymentReceived"
[1744074516479] ERROR: error occurred during block processing
    component: "container"
    indexer: "sep"
    blockNumber: 7802789
    err: {
      "type": "DatabaseError",
      "message": "insert into \"spaces\" (\"_indexer\", \"block_range\", \"id\", \"turbo_expiration\", \"turbo_expiration_date\") values ($1, int8range($2, NULL), $3, $4, $5) - value \"7000206660\" is out of range for type integer",
      "stack":
          error: insert into "spaces" ("_indexer", "block_range", "id", "turbo_expiration", "turbo_expiration_date") values ($1, int8range($2, NULL), $3, $4, $5) - value "7000206660" is out of range for type integer
              at Parser.parseErrorMessage (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:369:69)
              at Parser.handlePacket (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:188:21)
              at Parser.parse (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:103:30)
              at Socket.<anonymous> (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/index.ts:7:48)
              at Socket.emit (node:events:518:28)
              at Socket.emit (node:domain:489:12)
              at addChunk (node:internal/streams/readable:561:12)
              at readableAddChunkPushByteMode (node:internal/streams/readable:512:3)
              at Socket.Readable.push (node:internal/streams/readable:392:5)
              at TCP.onStreamRead (node:internal/stream_base_commons:189:23)
      "length": 148,
      "name": "error",
      "severity": "ERROR",
      "code": "22003",
      "where": "unnamed portal parameter $4 = '...'",
      "file": "numutils.c",
      "line": "317",
      "routine": "pg_strtoint32"
    }
In handlePaymentReceived
```

The data seems to be inserted in the db, as I see a row with this block number, but all subsequent import fails with

```
[1744074812708] INFO: found contract event
    component: "container"
    indexer: "sep"
    contract: "0xe40bfeb5a3014c9b98597088ca71eccdc27ca410"
    event: "PaymentReceived(address,address,uint256,string)"
    handlerFn: "handlePaymentReceived"
[1744074813095] ERROR: error occurred during block processing
    component: "container"
    indexer: "sep"
    blockNumber: 7802789
    err: {
      "type": "DatabaseError",
      "message": "insert into \"payments\" (\"_indexer\", \"amount_decimal\", \"amount_raw\", \"barcode\", \"block\", \"block_range\", \"id\", \"sender\", \"space\", \"timestamp\", \"token_address\", \"token_symbol\", \"type\") values ($1, $2, $3, $4, $5, int8range($6, NULL), $7, $8, $9, $10, $11, $12, $13) - conflicting key value violates exclusion constraint \"payments_id__indexer_block_range_excl\"",
      "stack":
          error: insert into "payments" ("_indexer", "amount_decimal", "amount_raw", "barcode", "block", "block_range", "id", "sender", "space", "timestamp", "token_address", "token_symbol", "type") values ($1, $2, $3, $4, $5, int8range($6, NULL), $7, $8, $9, $10, $11, $12, $13) - conflicting key value violates exclusion constraint "payments_id__indexer_block_range_excl"
              at Parser.parseErrorMessage (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:369:69)
              at Parser.handlePacket (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:188:21)
              at Parser.parse (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:103:30)
              at Socket.<anonymous> (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/index.ts:7:48)
              at Socket.emit (node:events:518:28)
              at Socket.emit (node:domain:489:12)
              at addChunk (node:internal/streams/readable:561:12)
              at readableAddChunkPushByteMode (node:internal/streams/readable:512:3)
              at Socket.Readable.push (node:internal/streams/readable:392:5)
              at TCP.onStreamRead (node:internal/stream_base_commons:189:23)
      "length": 497,
      "name": "error",
      "severity": "ERROR",
      "code": "23P01",
      "detail": "Key (id, _indexer, block_range)=(0xe6cd1f0c82d5095be50e45c3e14246527b47950d62ac357eaf4d6280044b3247, sep, [7802789,)) conflicts with existing key (id, _indexer, block_range)=(0xe6cd1f0c82d5095be50e45c3e14246527b47950d62ac357eaf4d6280044b3247, sep, [7802789,)).",
      "schema": "public",
      "table": "payments",
      "constraint": "payments_id__indexer_block_range_excl",
      "file": "execIndexing.c",
      "line": "856",
      "routine": "check_exclusion_or_unique_constraint"
    }
In handlePaymentReceived
```

wa0x6e commented on src/writers.ts:
```diff
@@ -1,12 +1,78 @@
+  }
+
+  // If the space already has an expiration date, use it as the current expiration date
+  const currentExpirationTimestamp =
+    space.turbo_expiration == 0 ? blockTimestamp : space.turbo_expiration;
```
```suggestion
  const currentExpirationTimestamp =
    space.turbo_expiration ||  blockTimestamp;
```
```suggestion
  const currentExpirationTimestamp =
    space.turbo_expiration == 0 ? blockTimestamp : space.turbo_expiration;
```

pscott replied:
does `||` work? won't it return a boolean instead ? I'm lost with JS lol

wa0x6e replied:
> does `||` work? won't it return a boolean instead ? I'm lost with JS lol

Yes

wa0x6e commented on src/writers.ts:
```diff
@@ -32,10 +98,28 @@ export function createEvmWriters(indexerName: string) {
+      space = new Space(metadata.params.space, indexerName);
+    }
+
+    const expirationDate = computeExpiration(space, payment, metadata, block.timestamp);
```
As `computeExpiration` is returning `new Date(0)` in case the sender has send less than one month amount, wouldn't this be set to timestamp 0, and moving back the expiration date for an existing space with a previously valid expiration date ?

pscott replied:
you are right. 
5b217bb6ef60f4a8cf8641f4c11c3ccd4e4b8978

pscott commented:
If a first-time user sends `0.001` USDC (not enough to increase turbo). Should we still create an `space` entry for the user?

pscott commented:
> I am getting this error
> 
> ```
> [1744074515668] INFO: found contract event
>     component: "container"
>     indexer: "sep"
>     contract: "0xe40bfeb5a3014c9b98597088ca71eccdc27ca410"
>     event: "PaymentReceived(address,address,uint256,string)"
>     handlerFn: "handlePaymentReceived"
> [1744074516479] ERROR: error occurred during block processing
>     component: "container"
>     indexer: "sep"
>     blockNumber: 7802789
>     err: {
>       "type": "DatabaseError",
>       "message": "insert into \"spaces\" (\"_indexer\", \"block_range\", \"id\", \"turbo_expiration\", \"turbo_expiration_date\") values ($1, int8range($2, NULL), $3, $4, $5) - value \"7000206660\" is out of range for type integer",
>       "stack":
>           error: insert into "spaces" ("_indexer", "block_range", "id", "turbo_expiration", "turbo_expiration_date") values ($1, int8range($2, NULL), $3, $4, $5) - value "7000206660" is out of range for type integer
>               at Parser.parseErrorMessage (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:369:69)
>               at Parser.handlePacket (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:188:21)
>               at Parser.parse (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:103:30)
>               at Socket.<anonymous> (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/index.ts:7:48)
>               at Socket.emit (node:events:518:28)
>               at Socket.emit (node:domain:489:12)
>               at addChunk (node:internal/streams/readable:561:12)
>               at readableAddChunkPushByteMode (node:internal/streams/readable:512:3)
>               at Socket.Readable.push (node:internal/streams/readable:392:5)
>               at TCP.onStreamRead (node:internal/stream_base_commons:189:23)
>       "length": 148,
>       "name": "error",
>       "severity": "ERROR",
>       "code": "22003",
>       "where": "unnamed portal parameter $4 = '...'",
>       "file": "numutils.c",
>       "line": "317",
>       "routine": "pg_strtoint32"
>     }
> In handlePaymentReceived
> ```
> 
> The data seems to be inserted in the db, as I see a row with this block number, but all subsequent import fails with
> 
> ```
> [1744074812708] INFO: found contract event
>     component: "container"
>     indexer: "sep"
>     contract: "0xe40bfeb5a3014c9b98597088ca71eccdc27ca410"
>     event: "PaymentReceived(address,address,uint256,string)"
>     handlerFn: "handlePaymentReceived"
> [1744074813095] ERROR: error occurred during block processing
>     component: "container"
>     indexer: "sep"
>     blockNumber: 7802789
>     err: {
>       "type": "DatabaseError",
>       "message": "insert into \"payments\" (\"_indexer\", \"amount_decimal\", \"amount_raw\", \"barcode\", \"block\", \"block_range\", \"id\", \"sender\", \"space\", \"timestamp\", \"token_address\", \"token_symbol\", \"type\") values ($1, $2, $3, $4, $5, int8range($6, NULL), $7, $8, $9, $10, $11, $12, $13) - conflicting key value violates exclusion constraint \"payments_id__indexer_block_range_excl\"",
>       "stack":
>           error: insert into "payments" ("_indexer", "amount_decimal", "amount_raw", "barcode", "block", "block_range", "id", "sender", "space", "timestamp", "token_address", "token_symbol", "type") values ($1, $2, $3, $4, $5, int8range($6, NULL), $7, $8, $9, $10, $11, $12, $13) - conflicting key value violates exclusion constraint "payments_id__indexer_block_range_excl"
>               at Parser.parseErrorMessage (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:369:69)
>               at Parser.handlePacket (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:188:21)
>               at Parser.parse (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:103:30)
>               at Socket.<anonymous> (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/index.ts:7:48)
>               at Socket.emit (node:events:518:28)
>               at Socket.emit (node:domain:489:12)
>               at addChunk (node:internal/streams/readable:561:12)
>               at readableAddChunkPushByteMode (node:internal/streams/readable:512:3)
>               at Socket.Readable.push (node:internal/streams/readable:392:5)
>               at TCP.onStreamRead (node:internal/stream_base_commons:189:23)
>       "length": 497,
>       "name": "error",
>       "severity": "ERROR",
>       "code": "23P01",
>       "detail": "Key (id, _indexer, block_range)=(0xe6cd1f0c82d5095be50e45c3e14246527b47950d62ac357eaf4d6280044b3247, sep, [7802789,)) conflicts with existing key (id, _indexer, block_range)=(0xe6cd1f0c82d5095be50e45c3e14246527b47950d62ac357eaf4d6280044b3247, sep, [7802789,)).",
>       "schema": "public",
>       "table": "payments",
>       "constraint": "payments_id__indexer_block_range_excl",
>       "file": "execIndexing.c",
>       "line": "856",
>       "routine": "check_exclusion_or_unique_constraint"
>     }
> In handlePaymentReceived
> ```

Came in when I moved to env vars, forgot to adjust using `DECIMALS`.
Fixed: b9b0efbd535c06b0b4c1ba99ff32f2f11b820a7f

wa0x6e commented:
Getting this error after a few blocks

```
[1744324665247] INFO: found contract event
    component: "container"
    indexer: "sep"
    contract: "0xe40bfeb5a3014c9b98597088ca71eccdc27ca410"
    event: "PaymentReceived(address,address,uint256,string)"
    handlerFn: "handlePaymentReceived"
[1744324666116] ERROR: error occurred during block processing
    component: "container"
    indexer: "sep"
    blockNumber: 8073476
    err: {
      "type": "DatabaseError",
      "message": "insert into \"payments\" (\"_indexer\", \"amount_decimal\", \"amount_raw\", \"barcode\", \"block\", \"block_range\", \"id\", \"sender\", \"space\", \"timestamp\", \"token_address\", \"token_symbol\", \"type\") values ($1, $2, $3, $4, $5, int8range($6, NULL), $7, $8, $9, $10, $11, $12, $13) - conflicting key value violates exclusion constraint \"payments_id__indexer_block_range_excl\"",
      "stack":
          error: insert into "payments" ("_indexer", "amount_decimal", "amount_raw", "barcode", "block", "block_range", "id", "sender", "space", "timestamp", "token_address", "token_symbol", "type") values ($1, $2, $3, $4, $5, int8range($6, NULL), $7, $8, $9, $10, $11, $12, $13) - conflicting key value violates exclusion constraint "payments_id__indexer_block_range_excl"
              at Parser.parseErrorMessage (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:369:69)
              at Parser.handlePacket (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:188:21)
              at Parser.parse (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:103:30)
              at Socket.<anonymous> (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/index.ts:7:48)
              at Socket.emit (node:events:518:28)
              at Socket.emit (node:domain:489:12)
              at addChunk (node:internal/streams/readable:561:12)
              at readableAddChunkPushByteMode (node:internal/streams/readable:512:3)
              at Socket.Readable.push (node:internal/streams/readable:392:5)
              at TCP.onStreamRead (node:internal/stream_base_commons:189:23)
      "length": 497,
      "name": "error",
      "severity": "ERROR",
      "code": "23P01",
      "detail": "Key (id, _indexer, block_range)=(0x6aba41fd72e60896bfc5aa212067992862c2b8e6ad95b376bacf003d84ceb997, sep, [8073476,)) conflicts with existing key (id, _indexer, block_range)=(0x6aba41fd72e60896bfc5aa212067992862c2b8e6ad95b376bacf003d84ceb997, sep, [8073476,)).",
      "schema": "public",
      "table": "payments",
      "constraint": "payments_id__indexer_block_range_excl",
      "file": "execIndexing.c",
      "line": "856",
      "routine": "check_exclusion_or_unique_constraint"
    }
In handlePaymentReceived
```

bonustrack commented on .env.example:
```diff
@@ -1 +1,4 @@
 DATABASE_URL=postgres://postgres:default_password@localhost:5432/checkpoint
+ADMIN_ADDRESS='0x8C28Cf33d9Fd3D0293f963b1cd27e3FF422B425c'
+TURBO_MONTHLY_PRICE=600
+TURBO_YEARLY_PRICE=6000
```
If at some point we need to change it we wont be able to do that by just changing the env vars, that would mess the duration of plans that was paid before price change. I believe we should hardcode these values instead, and later keep the history of pricing also in the code

pscott replied:
requested here: https://github.com/snapshot-labs/schnaps-api/pull/2#discussion_r2018790049

pscott replied:
do you want to go back to how it was before? i e hardcoded

wa0x6e replied:
Now that we have unlimited supply of tests tokens, my request is now obsolete

pscott replied:
d6a33d2e1a707afc60e16859513d133d7c64249a

wa0x6e commented on src/writers.ts:
```diff
@@ -1,12 +1,84 @@
+    const surplus = Number(payment.amount_raw) % TURBO_YEARLY_PRICE;
+    const surplusSeconds = surplus / YEARLY_PRICE_PER_SECOND;
+    expirationDate.setSeconds(expirationDate.getSeconds() + surplusSeconds);
+  } else if (payment.amount_raw >= TURBO_MONTHLY_PRICE) {
```
This can be converted to a simple `else`, since we're checking only 2 scenarios:

- `>=` 1 year
- `<` 1 year

and remove the `else` on line 75, since we're already using a guard on line 49

pscott replied:
not a big fan of removing this. The code explicitly states `unreachable code`, and if in the future we touch on the guard on line `49`, it will affect future lines, without it being explicit

wa0x6e commented on src/writers.ts:
```diff
@@ -1,12 +1,84 @@
+  }
+
+  // If the space already has an expiration date, use it as the current expiration date
+  const currentExpirationTimestamp =
```
This assume that subscription are continuous, and ignoring the case where a space has stopped paying for turbo for a while, then subscribe again, in which case it will just extend from the last expiration, instead of restarting from the payment date

pscott replied:
you are correct. 0dfb8ec612750d27a38e8d924638d24db7874d9e

pscott commented:
> Getting this error after a few blocks
> 
> ```
> [1744324665247] INFO: found contract event
>     component: "container"
>     indexer: "sep"
>     contract: "0xe40bfeb5a3014c9b98597088ca71eccdc27ca410"
>     event: "PaymentReceived(address,address,uint256,string)"
>     handlerFn: "handlePaymentReceived"
> [1744324666116] ERROR: error occurred during block processing
>     component: "container"
>     indexer: "sep"
>     blockNumber: 8073476
>     err: {
>       "type": "DatabaseError",
>       "message": "insert into \"payments\" (\"_indexer\", \"amount_decimal\", \"amount_raw\", \"barcode\", \"block\", \"block_range\", \"id\", \"sender\", \"space\", \"timestamp\", \"token_address\", \"token_symbol\", \"type\") values ($1, $2, $3, $4, $5, int8range($6, NULL), $7, $8, $9, $10, $11, $12, $13) - conflicting key value violates exclusion constraint \"payments_id__indexer_block_range_excl\"",
>       "stack":
>           error: insert into "payments" ("_indexer", "amount_decimal", "amount_raw", "barcode", "block", "block_range", "id", "sender", "space", "timestamp", "token_address", "token_symbol", "type") values ($1, $2, $3, $4, $5, int8range($6, NULL), $7, $8, $9, $10, $11, $12, $13) - conflicting key value violates exclusion constraint "payments_id__indexer_block_range_excl"
>               at Parser.parseErrorMessage (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:369:69)
>               at Parser.handlePacket (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:188:21)
>               at Parser.parse (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:103:30)
>               at Socket.<anonymous> (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/index.ts:7:48)
>               at Socket.emit (node:events:518:28)
>               at Socket.emit (node:domain:489:12)
>               at addChunk (node:internal/streams/readable:561:12)
>               at readableAddChunkPushByteMode (node:internal/streams/readable:512:3)
>               at Socket.Readable.push (node:internal/streams/readable:392:5)
>               at TCP.onStreamRead (node:internal/stream_base_commons:189:23)
>       "length": 497,
>       "name": "error",
>       "severity": "ERROR",
>       "code": "23P01",
>       "detail": "Key (id, _indexer, block_range)=(0x6aba41fd72e60896bfc5aa212067992862c2b8e6ad95b376bacf003d84ceb997, sep, [8073476,)) conflicts with existing key (id, _indexer, block_range)=(0x6aba41fd72e60896bfc5aa212067992862c2b8e6ad95b376bacf003d84ceb997, sep, [8073476,)).",
>       "schema": "public",
>       "table": "payments",
>       "constraint": "payments_id__indexer_block_range_excl",
>       "file": "execIndexing.c",
>       "line": "856",
>       "routine": "check_exclusion_or_unique_constraint"
>     }
> In handlePaymentReceived
> ```

I'm not getting this error. Which network are you indexing?

wa0x6e commented:
> > Getting this error after a few blocks
> > ```
> > [1744324665247] INFO: found contract event
> >     component: "container"
> >     indexer: "sep"
> >     contract: "0xe40bfeb5a3014c9b98597088ca71eccdc27ca410"
> >     event: "PaymentReceived(address,address,uint256,string)"
> >     handlerFn: "handlePaymentReceived"
> > [1744324666116] ERROR: error occurred during block processing
> >     component: "container"
> >     indexer: "sep"
> >     blockNumber: 8073476
> >     err: {
> >       "type": "DatabaseError",
> >       "message": "insert into \"payments\" (\"_indexer\", \"amount_decimal\", \"amount_raw\", \"barcode\", \"block\", \"block_range\", \"id\", \"sender\", \"space\", \"timestamp\", \"token_address\", \"token_symbol\", \"type\") values ($1, $2, $3, $4, $5, int8range($6, NULL), $7, $8, $9, $10, $11, $12, $13) - conflicting key value violates exclusion constraint \"payments_id__indexer_block_range_excl\"",
> >       "stack":
> >           error: insert into "payments" ("_indexer", "amount_decimal", "amount_raw", "barcode", "block", "block_range", "id", "sender", "space", "timestamp", "token_address", "token_symbol", "type") values ($1, $2, $3, $4, $5, int8range($6, NULL), $7, $8, $9, $10, $11, $12, $13) - conflicting key value violates exclusion constraint "payments_id__indexer_block_range_excl"
> >               at Parser.parseErrorMessage (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:369:69)
> >               at Parser.handlePacket (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:188:21)
> >               at Parser.parse (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/parser.ts:103:30)
> >               at Socket.<anonymous> (/Users/wan/Documents/GitHub/schnaps-api/node_modules/pg-protocol/src/index.ts:7:48)
> >               at Socket.emit (node:events:518:28)
> >               at Socket.emit (node:domain:489:12)
> >               at addChunk (node:internal/streams/readable:561:12)
> >               at readableAddChunkPushByteMode (node:internal/streams/readable:512:3)
> >               at Socket.Readable.push (node:internal/streams/readable:392:5)
> >               at TCP.onStreamRead (node:internal/stream_base_commons:189:23)
> >       "length": 497,
> >       "name": "error",
> >       "severity": "ERROR",
> >       "code": "23P01",
> >       "detail": "Key (id, _indexer, block_range)=(0x6aba41fd72e60896bfc5aa212067992862c2b8e6ad95b376bacf003d84ceb997, sep, [8073476,)) conflicts with existing key (id, _indexer, block_range)=(0x6aba41fd72e60896bfc5aa212067992862c2b8e6ad95b376bacf003d84ceb997, sep, [8073476,)).",
> >       "schema": "public",
> >       "table": "payments",
> >       "constraint": "payments_id__indexer_block_range_excl",
> >       "file": "execIndexing.c",
> >       "line": "856",
> >       "routine": "check_exclusion_or_unique_constraint"
> >     }
> > In handlePaymentReceived
> > ```
> 
> I'm not getting this error. Which network are you indexing?

sep, I start with `INDEX_TESTNET=1 yarn dev`

wa0x6e commented:
There's some date computation imprecision

Taking a 26 months subscription today (14/04/2025) is giving an expiration on 26/06/2027, with 12 extra days

wa0x6e commented:
You can test using https://feat-connect-turbo-to-shnaps-preview-snapshot-testnet.on-fleek.app/#/ 

Go to some spaces, and append /pro to the url

pscott commented:
> Taking a 26 months subscription today (14/04/2025) is giving an expiration on 26/06/2027, with 12 extra days

This is because you sent $13200, and the yearly price is 6k.
So you get two years ($12k).
You have $1200 left, which is two months ($1k).
You have $200 left, which is around 200/500 around 12 days.

This is *not* intuitive but is what @bonustrack requested.
I think a more intuitive approach (and the one you had in mind @wa0x6e also) would be: $13200 -> 2 years at yearly rate, and 2 months at monthly rate (12k + 1200).

wa0x6e commented on src/writers.ts:
```diff
@@ -1,21 +1,95 @@
+    }
+  }
+
+  // If the space already has an expiration date, use it as the current expiration date
+  let currentExpirationTimestamp = space.turbo_expiration || blockTimestamp;
+  if (currentExpirationTimestamp < blockTimestamp) {
+    // Case where a user discontinued the payment and then resumed it
+    currentExpirationTimestamp = blockTimestamp;
+  }
```
```suggestion
  // If the space already has an expiration date, use it as the current expiration date
  const currentExpirationTimestamp = Max.math(space.turbo_expiration, blockTimestamp);
```

wa0x6e replied:
sorry, typo, it's `Math.max`
